indra/newview/llhudeffectresetskeleton.cpp

```diff
@@ -0,0 +1,207 @@
+void LLHUDEffectResetSkeleton::unpackData(LLMessageSystem *mesgsys, S32 blocknum)
+{
+	LLVector3d new_target;
+	U8 packed_data[PKT_SIZE];
+
+	
+	LLHUDEffect::unpackData(mesgsys, blocknum);
+	
+	LLUUID source_id;
+	mesgsys->getUUIDFast(_PREHASH_Effect, _PREHASH_AgentID, source_id, blocknum);
+	
+	LLViewerObject *objp = gObjectList.findObject(source_id);
+	if (objp && objp->isAvatar())
+	{
+		setSourceObject(objp);
+	}
+	else
+	{
+		//LL_WARNS() << "Could not find source avatar for ResetSkeleton effect" << LL_ENDL;
+		return;
+	}
+	
+	S32 size = mesgsys->getSizeFast(_PREHASH_Effect, blocknum, _PREHASH_TypeData);
+	if (size != PKT_SIZE)
+	{
+		LL_WARNS() << "ResetSkeleton effect with bad size " << size << LL_ENDL;
+		return;
+	}
+	
+	mesgsys->getBinaryDataFast(_PREHASH_Effect, _PREHASH_TypeData, packed_data, PKT_SIZE, blocknum);
+	
+	LLUUID target_id;
+	htolememcpy(target_id.mData, &(packed_data[TARGET_OBJECT]), MVT_LLUUID, 16);
+
+	if (target_id.isNull())
+	{
+		target_id = source_id;
+	}
+	
+	objp = gObjectList.findObject(target_id);
+
+	if (objp)
+	{
+		setTargetObject(objp);
+	}
+	
+	U8 resetAnimations = 0;
+	htolememcpy(&resetAnimations, &(packed_data[RESET_ANIMATIONS]), MVT_U8, 1);
+	
+	// Pre-emptively assume this is going to be flags in the future.
+	// It isn't needed now, but this will assure that only bit 1 is set
+	mResetAnimations = resetAnimations & 1;
+	
+	update();
+}
+
+//-----------------------------------------------------------------------------
+// setTargetObjectAndOffset()
+//-----------------------------------------------------------------------------
+void LLHUDEffectResetSkeleton::setTargetObject(LLViewerObject *objp)
+{
+	mTargetObject = objp;
+}
+
+
+//-----------------------------------------------------------------------------
+// markDead()
+//-----------------------------------------------------------------------------
+void LLHUDEffectResetSkeleton::markDead()
+{
+	LLHUDEffect::markDead();
+}
+
+void LLHUDEffectResetSkeleton::setSourceObject(LLViewerObject* objectp)
+{
+	// restrict source objects to avatars
+	if (objectp && objectp->isAvatar())
+	{
+		LLHUDEffect::setSourceObject(objectp);
+	}
+}
+
+//-----------------------------------------------------------------------------
+// update()
+//-----------------------------------------------------------------------------
+void LLHUDEffectResetSkeleton::update()
+{
+	// If the target object is dead, set the target object to NULL
+	if (mTargetObject.isNull() || mTargetObject->isDead())
+	{
+		markDead();
+		return;
+	}
+
+	if (mSourceObject.isNull() || mSourceObject->isDead())
+	{
+		markDead();
+		return;
+	}
+	
+	bool owned = false;
+	if(mTargetObject->isAnimatedObject())
+	{
+		owned = mTargetObject->mOwnerID == mSourceObject->getID();
+	}
+	else
+	{
+		owned = mTargetObject->getID() == mSourceObject->getID();
+	}
+
+	if (owned)
+	{
+		if (mTargetObject->isAvatar() || mTargetObject->isAnimatedObject())
```

Comment:
This sems wrong. Animated objects have attached control avatars but aren't avatars so you need to mVObjp->getControlAvatar() and then reset skeleton of the control avatar.